### DISCUSSION TITLE
mature release of FastAPI[Question]

### DISCUSSION THREAD

**Discussion by neverset123** [fastapi/fastapi#8997] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

Hi,
I already used FastAPI for some time. But I still have a confusion in my mind:
Why until now (18.01.2021) there is still not a release with main version number? Is there some Issues that blocking the author to make a "real" mature release with version number higher than 1.0.0?

Thanks for all feedbacks!

**Comment by ycd** [comment]
Are there any missing features that you expect to see in a "mature" framework?

**Comment by neverset123** [comment]
> Are there any missing features that you expect to see in a "mature" framework?

until now I think fastAPI is quite good. according to semantic versioning  (Major.Minor.Patch), "The presence of a pre-release tag (-alpha, -beta) indicates substantial risk, as does a major number of zero (0.y.z), which is used to indicate a work-in-progress that may contain any level of potentially breaking changes (highest risk)", citied from wiki

**Comment by categulario** [comment]
I think there are two things that would be very important:

1. a major release, with stability warranties.
2. API docs. Although the tutorials are very extensive, having API docs would be very helpful for people that want build things that are not even considered in the tutorials (api-token comes to mind).

right now having no 1.0.0 feels like things might break at some point (for instance, in the 1.0.0 release) like it happens every once in a while in hugo static site generator.

**Comment by ycd** [comment]
Fair points, but i don't think none of the possible changes in the future will be hard to switch to.

`FastAPI` has a slightly simple syntax, on the other hand we are dependant to two other libraries.
 1. `pydantic`
 2. `starlette`

As you can see from our `__init__.py` file most of our functions are re-exported from `Starlette`

https://github.com/tiangolo/fastapi/blob/41f291524dd16575f01505f09bb4a85e98676dab/fastapi/__init__.py#L7-L24

This means some of the changes will not be dependant to `FastAPI`, like it happened it when `Starlette` deprecated `UJSONResponse`.

If you take a look at our [Release Notes](https://fastapi.tiangolo.com/release-notes/), you 'll find out we had 4 breaking change.

 1. #463 - **Updated pydantic to 0.32.2.**
 1. #1862 - **Removed support for deprecated Pydantic 0.32.2.**
 1. #2434 - **Added support for shared/top-level parameters-dependencies**
 1. #2548 - **Upgraded Uvicorn version when installing with `all` option**

As a slightly young framework, we had one "real" breaking change, #2434. Even in that change we deleted more code than we wrote, because it helped us to delete code repetitions.

Most of FastAPI's internal part are wrappers around `Starlette`, a change to Starlette will not break your code, because FastAPI wraps it up and gives you a more friendly API interface. 

That being said, these are my personal thoughts and from my perspective there is nothing to worry about.

**Comment by neverset123** [comment]
we cannot simply say FastAPI is young(although it is comparatively younger than other frameworks). Until now FastAPI has undergone more than 2 years and made 103 Releases, but all of them are iterations on minor or patch versioning. Compared to Tornado, the first release is stable release 1.0.0 one year after appearance. There must be some thoughts under @tiangolo 's mind that he want to keep it as it is. Could you point us out on this confusion? Thanks!

**Comment by Kludex** [comment]
I can't find the interview he talked about it :cry: Some interview between July and November, a podcast or YouTube video has that answers that...

I'm going to say the way I remember (correct me if I'm wrong @tiangolo). The way I recall, he said that he wants to fix some too complex functions that we currently have, add some features from the backlog and add API documentation (which should be in the backlog). But really, you should wait for his words, I'm just expressing what I **think** I remember, it was too long ago. 

Maybe the admin interface based on OpenAPI was one of the features prerequisite for 1.0, but I'm not sure.

**Comment by FalseDev** [comment]
I think a better title would be `Stable release of FastAPI`

**Comment by mdgilene** [comment]
I'd like to add that not having a stable `1.0.0` release of FastAPI makes it difficult to include FastAPI as a dependency to other libraries. If you are using tools like poetry to manage dependency versions you end up locking your users into very specific FastAPI versions when there isn't a major release. For example:

Util Library
  -> depends on FastAPI ^0.65
  
Application
  -> depends on Util Library

Your application is now locked into 0.65 >= FastAPI <= 0.66 which is pretty limited.

**Comment by jykae** [comment]
Include support for API versioning out-of-the-box, like https://github.com/DeanWay/fastapi-versioning
